tensorflow/security/fuzzing/core/function/runtime_client_fuzz.cc

```diff
@@ -0,0 +1,242 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include <gtest/gtest.h>
+#include "fuzztest/fuzztest.h"
+#include "tensorflow/core/function/runtime_client.h"
+#include "tensorflow/core/ir/dialect.h"
+#include "tensorflow/core/ir/ops.h"
+#include "mlir/Parser/Parser.h"  // from @llvm-project
+#include "mlir/Pass/PassManager.h"  // from @llvm-project
+#include "mlir/Pass/Pass.h"  // from @llvm-project
+
+namespace tensorflow {
+namespace fuzzing {
+
+using namespace core::function;
+
+EagerContextPtr LocalEagerCtx() {
+  SessionOptions opts;
+  std::vector<std::unique_ptr<Device>> devices;
+  Status&& device_init_status = DeviceFactory::AddDevices(
+      opts, "/job:localhost/replica:0/task:0", &devices);
+  CHECK(device_init_status.ok());  // Crash OK
+
+  return EagerContextPtr(new EagerContext(
+      opts, ContextDevicePlacementPolicy::DEVICE_PLACEMENT_SILENT,
+      /*async=*/false,
+      /*device_mgr=*/new DynamicDeviceMgr(std::move(devices)),
+      /*device_mgr_owned=*/true,
+      /*rendezvous=*/nullptr,
+      /*cluster_flr=*/nullptr,
+      /*collective_executor_mgr=*/nullptr,
+      /*run_eager_op_as_function=*/true));
+}
+
+FunctionDef EmptyFunctionDefGenerator(int number_of_input_arguments, int number_of_output_arguments) {
+  std::vector<string> in_def_vec;
+  in_def_vec.reserve(number_of_input_arguments);
+  for (int c = 0; c < number_of_input_arguments; ++c) {
+    in_def_vec.push_back("in" + std::to_string(c) + ":float");
+  }
+  std::vector<FunctionDefHelper::Node> body_nodes;
+  if (number_of_output_arguments > number_of_input_arguments) {
+    Tensor const_value(DataTypeToEnum<float>::value, {});
+    const_value.scalar<float>()() = 0;
+    body_nodes.push_back({{"zero"}, "Const", {}, {{"value", const_value}, {"dtype", DT_FLOAT}}});
+  }
+  std::vector<string> out_def_vec;
+  out_def_vec.reserve(number_of_output_arguments);
+  std::vector<std::pair<string, string>> ret_def;
+  ret_def.reserve(number_of_output_arguments);
+  for (int c = 0; c < number_of_output_arguments; ++c) {
+    auto output_id = "out" + std::to_string(c);
+    out_def_vec.push_back(output_id + ":float");
+    if (c < number_of_input_arguments) {
+      ret_def.emplace_back(output_id, "in" + std::to_string(c));
+    } else {
+      ret_def.emplace_back(output_id, "zero:output");
+    }
+  }
+  return FunctionDefHelper::Create("TestFunction", in_def_vec, out_def_vec, {}, body_nodes, ret_def);
+}
+
+auto NumberOfInputArguments() {
+  return fuzztest::InRange(0, 7);
+}
+
+auto NumberOfOutputArguments() {
+  return fuzztest::InRange(1, 7);
+}
+
+auto FunctionDefDomain() {
+  return fuzztest::Map(EmptyFunctionDefGenerator, NumberOfInputArguments(), NumberOfOutputArguments());
+}
+
+void CreateFunctionFuzz(FunctionDef def) {
+  auto ctx = LocalEagerCtx();
+  Runtime rt(*ctx);
```

Comment:
is Runtime creation time-consuming operation?
if so maybe there is sense to move it to the test fixture?

Reply:
The context here holds the current state. That state is denote if the function we create is present in the context or isn't. So the creation of the context is part of the test, and so is the creation of the Runtime.